File: front/src/lib/schemas/inference_batches.ts
"use client";

import { z } from "zod";

import { ClipSchema } from "./clips";
import { CustomModelSchema } from "./custom_models";

// Inference batch status enum
export const InferenceBatchStatusSchema = z.enum([
  "pending",
  "running",
  "completed",
  "failed",
  "cancelled",
]);

export type InferenceBatchStatus = z.infer<typeof InferenceBatchStatusSchema>;

// Main InferenceBatch schema
export const InferenceBatchSchema = z.object({
  uuid: z.string().uuid(),
  name: z.string().nullable(),
  ml_project_uuid: z.string().uuid(),
  custom_model: CustomModelSchema.nullable(),
  status: InferenceBatchStatusSchema,
  confidence_threshold: z.number(),
  total_clips: z.number().int(),
  processed_clips: z.number().int(),
  total_predictions: z.number().int(),
  positive_predictions_count: z.number().int(),
  negative_predictions_count: z.number().int(),
  average_confidence: z.number().nullable(),
  started_at: z.coerce.date().nullable(),
  completed_at: z.coerce.date().nullable(),
  duration_seconds: z.number().nullable(),
  error_message: z.string().nullable(),
  description: z.string().nullable(),
  created_by_id: z.string().uuid(),
  created_on: z.coerce.date(),
});

export type InferenceBatch = z.infer<typeof InferenceBatchSchema>;

// Create schema
export const InferenceBatchCreateSchema = z.object({
  name: z.string().optional(),
  custom_model_id: z.number().int().optional(),
  custom_model_uuid: z.string().uuid().optional(),
  confidence_threshold: z.number().min(0).max(1).optional(),
  clip_ids: z.array(z.number().int()).optional(),
  include_all_clips: z.boolean().optional(),
  exclude_already_labeled: z.boolean().optional(),
  description: z.string().optional(),
});

export type InferenceBatchCreate = z.infer<typeof InferenceBatchCreateSchema>;

// InferencePrediction schema
export const InferencePredictionSchema = z.object({
  uuid: z.string().uuid(),
  inference_batch_uuid: z.string().uuid(),
  clip: ClipSchema,
  confidence: z.number(),
  predicted_positive: z.boolean(),
  created_on: z.coerce.date(),
});

export type InferencePrediction = z.infer<typeof InferencePredictionSchema>;

// Inference progress schema
export const InferenceProgressSchema = z.object({
  status: InferenceBatchStatusSchema,
  progress: z.number().min(0).max(100),
  processed_items: z.number().int(),
  total_items: z.number().int(),
  positive_predictions: z.number().int(),
});

export type InferenceProgress = z.infer<typeof InferenceProgressSchema>;

// Convert to annotation project request schema
export const ConvertToAnnotationProjectRequestSchema = z.object({
  name: z.string().min(1).max(255),
  description: z.string().optional(),
  confidence_threshold: z.number().min(0).max(1).optional(),
  include_only_positive: z.boolean().optional().default(true),
});

export type ConvertToAnnotationProjectRequest = z.infer<
  typeof ConvertToAnnotationProjectRequestSchema
>;

// Note: ConvertToAnnotationProjectResponse is not needed as the API returns AnnotationProject

File: old/back/src/echoroo/schemas/inference_batches.py
"""Schemas for Inference Batches and Predictions."""

import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, Field

from echoroo.schemas.base import BaseSchema
from echoroo.schemas.clips import Clip
from echoroo.schemas.custom_models import CustomModel

__all__ = [
    "InferenceBatchStatus",
    "InferenceBatch",
    "InferenceBatchCreate",
    "InferencePrediction",
    "InferenceProgress",
    "InferenceBatchStats",
    "ConvertToAnnotationProjectRequest",
]


class InferenceBatchStatus(str, Enum):
    """Status of an inference batch."""

    PENDING = "pending"
    """Batch is queued for processing."""

    RUNNING = "running"
    """Inference is currently running."""

    COMPLETED = "completed"
    """Inference completed successfully."""

    FAILED = "failed"
    """Inference failed due to an error."""

    CANCELLED = "cancelled"
    """Inference was cancelled by the user."""


class InferenceBatchCreate(BaseModel):
    """Schema for creating an inference batch."""

    name: str | None = Field(default=None, max_length=255)
    """Optional name for the inference batch."""

    custom_model_id: int | None = Field(
        default=None,
        description="Custom model ID to use for inference",
    )
    """Custom model ID to use for running inference."""

    custom_model_uuid: UUID | None = Field(
        default=None,
        description="Custom model UUID to use for inference (alternative to custom_model_id)",
    )
    """Custom model UUID to use for running inference."""

    confidence_threshold: float = Field(
        default=0.5,
        ge=0.0,
        le=1.0,
        description="Minimum confidence for predictions",
    )
    """Minimum confidence score to include predictions."""

    clip_ids: list[int] | None = Field(
        default=None,
        description="Specific clips to run inference on",
    )
    """Optional list of specific clip IDs to process."""

    include_all_clips: bool = Field(
        default=False,
        description="Run inference on all clips in the dataset",
    )
    """Whether to run inference on all clips in the dataset."""

    exclude_already_labeled: bool = Field(
        default=True,
        description="Skip clips that already have labels",
    )
    """Whether to skip clips that already have labels from search sessions."""

    description: str | None = Field(default=None, max_length=2000)
    """Optional description of the inference batch."""


class InferencePrediction(BaseSchema):
    """Schema for an inference prediction returned to the user."""

    uuid: UUID
    """UUID of the prediction."""

    id: int = Field(..., exclude=True)
    """Database ID of the prediction."""

    inference_batch_id: int = Field(..., exclude=True)
    """Inference batch that produced this prediction."""

    inference_batch_uuid: UUID
    """UUID of the owning inference batch."""

    clip_id: int = Field(..., exclude=True)
    """Clip identifier for this prediction."""

    clip: Clip
    """Hydrated clip information."""

    confidence: float = Field(ge=0.0, le=1.0)
    """Confidence score for the prediction."""

    predicted_positive: bool
    """Whether the model predicted positive for target sound."""


class InferenceBatch(BaseSchema):
    """Schema for an inference batch returned to the user."""

    uuid: UUID
    """UUID of the inference batch."""

    id: int = Field(..., exclude=True)
    """Database ID of the inference batch."""

    name: str | None = None
    """Optional name for the inference batch."""

    ml_project_id: int = Field(..., exclude=True)
    """ML project that owns this batch."""

    ml_project_uuid: UUID
    """UUID of the owning ML project."""

    custom_model_id: int = Field(..., exclude=True)
    """Custom model identifier used for inference."""

    custom_model: CustomModel | None = None
    """Hydrated custom model information."""

    status: InferenceBatchStatus = InferenceBatchStatus.PENDING
    """Current batch status."""

    confidence_threshold: float
    """Confidence threshold used for predictions."""

    total_clips: int = 0
    """Total number of clips to process."""

    processed_clips: int = 0
    """Number of clips processed so far."""

    total_predictions: int = 0
    """Total number of predictions generated."""

    positive_predictions_count: int = 0
    """Number of predictions with positive label."""

    negative_predictions_count: int = 0
    """Number of predictions with negative label."""

    average_confidence: float | None = None
    """Average confidence score across all predictions (0.0 to 1.0)."""

    started_at: datetime.datetime | None = None
    """Timestamp when inference started."""

    completed_at: datetime.datetime | None = None
    """Timestamp when inference completed."""

    duration_seconds: float | None = None
    """Total inference duration in seconds."""

    error_message: str | None = None
    """Error message if inference failed."""

    description: str | None = None
    """Optional description of the inference batch."""

    created_by_id: UUID
    """User who created the inference batch."""


class InferenceProgress(BaseModel):
    """Schema for tracking inference progress."""

    status: InferenceBatchStatus
    """Current batch status."""

    total_clips: int = 0
    """Total number of clips to process."""

    processed_clips: int = 0
    """Number of clips processed so far."""

    predictions_generated: int = 0
    """Number of predictions generated so far."""

    progress_percent: float = 0.0
    """Processing progress percentage."""

    clips_per_second: float | None = None
    """Processing speed in clips per second."""

    estimated_time_remaining_seconds: float | None = None
    """Estimated time remaining for inference."""

    message: str | None = None
    """Human-readable status message."""


class InferenceBatchStats(BaseModel):
    """Aggregate statistics for an inference batch."""

    total_predictions: int = 0
    """Total number of predictions."""

    predictions_by_confidence: dict[str, int] = Field(default_factory=dict)
    """Count of predictions by confidence range (e.g., '0.9-1.0': 150)."""

    confidence_histogram: list[int] = Field(default_factory=list)
    """Histogram of confidence scores (10 buckets, 0.0-1.0)."""

    average_confidence: float | None = None
    """Average confidence score across all predictions."""

    median_confidence: float | None = None
    """Median confidence score."""


class ConvertToAnnotationProjectRequest(BaseModel):
    """Request payload for converting inference batch to annotation project."""

    name: str = Field(..., min_length=1, max_length=255)
    """Name of the annotation project to create."""

    description: str | None = None
    """Optional description of the annotation project."""

    confidence_threshold: float | None = Field(
        default=None,
        ge=0.0,
        le=1.0,
        description="Override confidence threshold (uses batch threshold if not provided)",
    )
    """Optional override for confidence threshold."""

    include_only_positive: bool = Field(
        default=True,
        description="Only include predictions where predicted_positive is True",
    )
    """Whether to filter to only positive predictions."""


class ConvertToAnnotationProjectResponse(BaseModel):
    """Response from converting inference batch to annotation project."""

    annotation_project_uuid: UUID
    """UUID of the created annotation project."""

    annotation_project_name: str
    """Name of the created annotation project."""

    total_tasks_created: int
    """Total number of annotation tasks created."""

    total_annotations_created: int
    """Total number of clip annotations created."""

    total_tags_added: int = 1
    """Total number of species tags added to the project."""
